addons/pkg/util/tkr_util_test.go

```diff
@@ -138,14 +142,74 @@ var _ = Describe("TKR utils", func() {
 			BeforeEach(func() {
 				crtCtl = &fakeclusterclient.CRTClusterClient{}
 				tkrName = testTKR
-				tkrV1Alpha3, err = GetTKRByNameV1Alpha3(ctx, crtCtl, tkrName)
+				tkrV1Alpha3, err = GetTKRByNameV1Alpha3(ctx, crtCtl, nil, tkrName)
 			})
 
 			It("should return nil TKR", func() {
 				Expect(err).ShouldNot(HaveOccurred())
 				Expect(tkrV1Alpha3).ShouldNot(BeNil())
 			})
 		})
+
+		When("tkr spec is available as annotations on cluster resource", func() {
+			BeforeEach(func() {
+				crtCtl = &fakeclusterclient.CRTClusterClient{}
+				tkrName = testTKR
+				var buf bytes.Buffer
+				out, _ := yaml.Marshal(tkrV1Alpha3)
+				w, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+				_, err = w.Write(out)
+				w.Close()
+				tkrString := base64.StdEncoding.EncodeToString(buf.Bytes())
+				annotation := make(map[string]string)
+				annotation["run.tanzu.vmware.com/tkr-spec"] = tkrString
+				cluster := &clusterapiv1beta1.Cluster{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Cluster",
+						APIVersion: "v1beta1",
+					},
+					ObjectMeta: metav1.ObjectMeta{Name: testClusterName, Namespace: testNamespace, Labels: map[string]string{}, Annotations: annotation},
+				}
+				tkrV1Alpha3, err = FetchTKRSpecFromAnnotations(cluster.Annotations)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(tkrV1Alpha3).NotTo(BeNil())
+			})
+
+			It("should return non nil TKR", func() {
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(tkrV1Alpha3).ShouldNot(BeNil())
+			})
+		})
+
+		When("tkr is not available calling GetTKRByNameV1Alpha3()", func() {
+			BeforeEach(func() {
+				crtCtl = &fakeclusterclient.CRTClusterClient{}
+				tkrName = testTKR
+				var buf bytes.Buffer
+				out, _ := yaml.Marshal(tkrV1Alpha3)
+				w, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+				_, err = w.Write(out)
+				w.Close()
+				tkrString := base64.StdEncoding.EncodeToString(buf.Bytes())
+				annotation := make(map[string]string)
+				annotation["run.tanzu.vmware.com/tkr-spec"] = tkrString
+				cluster := &clusterapiv1beta1.Cluster{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "Cluster",
+						APIVersion: "v1beta1",
+					},
+					ObjectMeta: metav1.ObjectMeta{Name: testClusterName, Namespace: testNamespace, Labels: map[string]string{}, Annotations: annotation},
+				}
+				tkrName = testTKR
+				crtCtl.GetReturns(apierrors.NewNotFound(schema.GroupResource{Resource: "TanzuKubernetesRelease"}, testTKR))
+				tkrV1Alpha3, err = GetTKRByNameV1Alpha3(ctx, crtCtl, cluster, tkrName)
+			})
+
+			It("should return non nil TKR", func() {
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(tkrV1Alpha3).ShouldNot(BeNil())
+			})
+		})
```

Comment:
you need to add the tests here covering the same but with cluster ==nil

Reply:
given that cluster resource will never be nil and the check is removed, the test for cluster==nil seems irrelevant now